src/cluster/connectionBuilder.js

```diff
@@ -1,10 +1,9 @@
 const Connection = require('../network/connection')
-const { KafkaJSNonRetriableError } = require('../errors')
-const shuffle = require('../utils/shuffle')
+const { KafkaJSConnectionError } = require('../errors')
 
 const validateBrokers = brokers => {
   if (!brokers || brokers.length === 0) {
-    throw new KafkaJSNonRetriableError(`Failed to connect: expected brokers array and got nothing`)
+    throw new KafkaJSConnectionError(`Failed to connect: expected brokers array and got nothing`)
```

Comment:
I don't think this should be a `KafkaJSConnectionError`, since it's actually a validation error that's not retriable. `KafkaJSConnectionError` is retriable by default. I see three options:

1. Keep it retriable and live with the fact that it's weird to validate the same static list N times.
2. Make `validateBrokers` a bit smarter and have it throw a retriable error if `brokers` is a function and a non-retriable error if it's not.
3. Wrap `brokers` in an async function that does retrying internally instead. Something along the lines of

```js
const retrier = createRetry({ ...retry })
const getBrokers = () => retrier(async bail => {
  const static = Array.isArray(brokers)
  const list = static ? brokers : await brokers()
  try {
    validateBrokers(list) // Throws retriable error
    return list
  } catch (err) {
    if (static) { // if it's a static list, don't bother retrying
      bail(err)
    } else {
      throw err
    }
  }
})
```

Reply:
I suggest second option

Reply:
Sounds good!